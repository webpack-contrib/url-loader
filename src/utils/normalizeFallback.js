import loaderUtils from 'loader-utils';

export default function normalizeFallback(fallback, originalOptions) {
  let loader = 'file-loader';
  let options = {};

  if (typeof fallback === 'string') {
    loader = fallback;

    const index = fallback.indexOf('?');

    if (index >= 0) {
      loader = fallback.substr(0, index);
      options = loaderUtils.parseQuery(fallback.substr(index));
    }
  }

  if (fallback !== null && typeof fallback === 'object') {
    ({ loader, options } = fallback);
  }

  options = Object.assign({}, originalOptions, options);

  delete options.fallback;

  return { loader, options };
}
