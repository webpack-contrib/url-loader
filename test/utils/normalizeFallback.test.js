/* eslint-disable
  prefer-destructuring,
*/
import normalizeFallback from '../../src/utils/normalizeFallback';

describe('normalizeFallback', () => {
  test('undefined', () => {
    // eslint-disable-next-line no-undefined
    const result = normalizeFallback(undefined, {
      limit: 8192,
      name: 'name-for-url-loader.[ext]',
    });

    expect(result).toMatchSnapshot();
  });

  test('string', () => {
    const result = normalizeFallback('file-loader', {
      limit: 8192,
      name: 'name-for-url-loader.[ext]',
    });

    expect(result).toMatchSnapshot();
  });

  test('string-with-query', () => {
    const result = normalizeFallback(
      'file-loader?name=name-for-file-loader.[ext]',
      { limit: 8192, name: 'name-for-url-loader.[ext]' }
    );

    expect(result).toMatchSnapshot();
  });

  test('object', () => {
    const result = normalizeFallback(
      { loader: 'file-loader' },
      { limit: 8192, name: 'name-for-url-loader.[ext]' }
    );

    expect(result).toMatchSnapshot();
  });

  test('object-with-options', () => {
    const result = normalizeFallback(
      {
        loader: 'file-loader',
        options: { name: 'name-for-file-loader.[ext]' },
      },
      { limit: 8192, name: 'name-for-url-loader.[ext]' }
    );

    expect(result).toMatchSnapshot();
  });
});
