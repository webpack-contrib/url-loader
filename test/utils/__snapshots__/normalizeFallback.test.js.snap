// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeFallback object 1`] = `
Object {
  "loader": "file-loader",
  "options": Object {
    "limit": 8192,
    "name": "name-for-url-loader.[ext]",
  },
}
`;

exports[`normalizeFallback object-with-options 1`] = `
Object {
  "loader": "file-loader",
  "options": Object {
    "limit": 8192,
    "name": "name-for-file-loader.[ext]",
  },
}
`;

exports[`normalizeFallback string 1`] = `
Object {
  "loader": "file-loader",
  "options": Object {
    "limit": 8192,
    "name": "name-for-url-loader.[ext]",
  },
}
`;

exports[`normalizeFallback string-with-query 1`] = `
Object {
  "loader": "file-loader",
  "options": Object {
    "limit": 8192,
    "name": "name-for-file-loader.[ext]",
  },
}
`;

exports[`normalizeFallback undefined 1`] = `
Object {
  "loader": "file-loader",
  "options": Object {
    "limit": 8192,
    "name": "name-for-url-loader.[ext]",
  },
}
`;
